### PR TITLE
Optimizations of `full_str_int`

### PR DESCRIPTION
**Pull request by vhavlena**
This PR contains:
- special axiom instantiation
- new preprocessing rules

### CODE REVIEW

**Comment by vhavlena** [comment]
On `full_str_int`:
```
# of formulae: 16968
###################################################################################
####                                   Table 1                                 ####
###################################################################################
| method                     |    max |      mean |   median |   std. dev |   TO+MO+ERR |   unknowns |
|----------------------------|--------|-----------|----------|------------|-------------|------------|
| z3-noodler-bdb7f83-2cddb2f |  84.32 | 0.225197  |     0.01 |   1.77001  |         715 |        895 |
| cvc5-1.0.8                 |  82.08 | 0.309905  |     0.03 |   1.18924  |           0 |          0 |
| z3-4.12.2                  | 118.32 | 1.44806   |     0.04 |   7.81162  |         234 |          0 |
| z3-noodler-1174fd1-2cddb2f |  96.78 | 0.130747  |     0.01 |   1.59691  |        1212 |       1199 |
```

File: src/smt/theory_str_noodler/var_union_find.h [hunk 1]
```diff
@@ -80,13 +80,26 @@ namespace smt::noodler {
...
+                int len = -1;
+                rational val;
+                if(this->m_util_a.is_numeral(t.m_key, val)) {
+                    len = val.get_int32();
+                }
                 for (const auto& s : t.m_value) {
                     std::string var = to_app(s)->get_decl()->get_name().str();
                     BasicTerm bvar(BasicTermType::Variable, var);
+                    
+                    if(len != -1 && len > 1) {
```
[the remark this answers is not in the document]
Comment (vhavlena) [reply src/smt/theory_str_noodler/var_union_find.h:96]: Well, it could be simplified, but yes, it is intentional.

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 2]
```diff
@@ -1207,11 +1208,33 @@ namespace smt::noodler {
             }
             expr_ref substr_in(m_util_s.re.mk_in_re(v, substr_re), m);
 
+            string_theory_propagation(xey);
             // 0 <= i <= |s| && 0 <= l <= |s| - i -> |v| in substr_re
             add_axiom({~i_ge_0, ~ls_le_i, ~l_ge_zero, ~li_ge_ls, mk_eq(le, l, false)});
             add_axiom({~i_ge_0, ~ls_le_i, ~l_ge_zero, ~li_ge_ls, mk_literal(substr_in)});
             // 0 <= i <= |s| && |s| < l + i  -> s = x.v
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:1215]: Is this comment still correct?

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 3]
```diff
@@ -1207,11 +1208,33 @@ namespace smt::noodler {
...
-            add_axiom({~i_ge_0, ~ls_le_i, li_ge_ls, mk_eq(xe, s, false)});
+            add_axiom({~i_ge_0, ~ls_le_i, li_ge_ls, mk_eq(y, eps, false)});
+            // 0 <= i <= |s| && l < 0 -> v = eps
+            add_axiom({~i_ge_0, ~ls_le_i, l_ge_zero, mk_eq(v, eps, false)});
+            // 0 <= i <= |s| -> xvy = s
+            add_axiom({~i_ge_0, ~ls_le_i, mk_eq(xey, s, false)});
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:1220]: Why is there `xvy` in the comment but `xey` in the code?

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 4]
```diff
@@ -1207,11 +1208,33 @@ namespace smt::noodler {
...
+            // i < 0 -> v = eps
+            add_axiom({i_ge_0, mk_eq(v, eps, false)});
+            // i > |s| -> v = eps
+            add_axiom({~ls_le_0, mk_eq(v, eps, false)});
+            add_axiom({ls_le_i, mk_eq(v, eps, false)});
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:1225]: Is this comment correct? Seems there is more happening in the code.

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 5]
```diff
@@ -1796,6 +1828,16 @@ namespace smt::noodler {
...
 
+        expr * sub_str = nullptr, *sub_ind = nullptr, *sub_len = nullptr;
+        rational val;
+        zstring str;
+        // handle the special case of the form (str.prefix "a" (str.substr s 5 2)) --> (str.at s 5) == "a"
+        if(m_util_s.str.is_string(x, str) && str.length() == 1 && m_util_s.str.is_extract(y, sub_str, sub_ind, sub_len) && m_util_a.is_numeral(sub_ind) && m_util_a.is_numeral(sub_len, val) && val.get_int32() >= 1) {
+            add_axiom({~mk_eq(x, m_util_s.str.mk_at(sub_str, sub_ind), false), mk_literal(e) });
+            add_axiom({mk_eq(x, m_util_s.str.mk_at(sub_str, sub_ind), false), ~mk_literal(e) });
+            return;
+        }
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:1839]: Shouldn't this be a rewriter rule?

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 6]
```diff
@@ -2016,21 +2058,32 @@ namespace smt::noodler {
...
         expr * ind = nullptr;
+        zstring str;
         if(expr_cases::is_contains_index(e, ind, m, m_util_s, m_util_a)) {
             expr_ref ind_eq(m.mk_eq( ind, m_util_a.mk_int(-1) ), m);
             expr_ref ind_leq(m_util_a.mk_le( ind, m_util_a.mk_int(-1) ), m);
             literal not_e = mk_literal(mk_not({e, m}));
             add_axiom({~mk_eq(ind, m_util_a.mk_int(-1), false), ~mk_literal(e) });
             add_axiom({mk_eq(ind, m_util_a.mk_int(-1), false), mk_literal(e) });
             return;
+        // if constains is of the form (str.constains str (str.at )) rewrite to a regular constaint
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:2075]:
```suggestion
        // if constains is of the form (str.constains strX (str.at ...)) rewrite to a regular constaint ((str.at ...) \in union of chars of strX)
```

This is what is happening here, right?

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 7]
```diff
@@ -2016,21 +2058,32 @@ namespace smt::noodler {
...
+        } else if (m_util_s.str.is_at(y) && m_util_s.str.is_string(x, str) && str.length() > 0) {
+            expr_ref re(m_util_s.re.mk_to_re(m_util_s.str.mk_string("")), m);
+            for(size_t i = 0; i < str.length(); i++) {
+                re = m_util_s.re.mk_union(re, m_util_s.re.mk_to_re(m_util_s.str.mk_string(str[i])));
+            }
+            expr_ref in_re(m_util_s.re.mk_in_re(y, re), m);
+            literal not_e = mk_literal(mk_not({e, m}));
+            add_axiom({not_e, mk_literal(in_re)});
+            return;
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:2084]: Shouldn't this be a rewrite rule?

File: src/smt/theory_str_noodler/formula_preprocess.cpp [hunk 8]
```diff
@@ -1335,12 +1337,76 @@ namespace smt::noodler {
...
+                    Predicate new_pred = Predicate(PredicateType::Equation, { Concat{c1[i]}, Concat(c2.begin(), c2.begin() + j + 1) });
+                    this->formula.add_predicate(new_pred);
+                    if(new_pred.get_right_side().size() > 1) {
+                        rem_ids.insert(pr2.first);
+                    }
+                } else if (j == 0) {
+                    Predicate new_pred = Predicate(PredicateType::Equation, { Concat{c2[j]}, Concat(c1.begin(), c1.begin() + i + 1) });
+                    this->formula.add_predicate(new_pred);
+                    if(new_pred.get_right_side().size() > 1) {
+                        rem_ids.insert(pr1.first);
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/formula_preprocess.cpp:1402]: Why is this removed?

**Review by jurajsic — APPROVED** [review]
Looks good, just some random questions

File: src/smt/theory_str_noodler/var_union_find.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (Adda0) [reply src/smt/theory_str_noodler/var_union_find.h:96]: Good. I was not sure whether this was a bug in the condition logic or not.

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 2 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:1216]: Yes the comments are bit messy, I have fixed them.

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 3 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp]: Fixed

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 5 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp]: I think so. I added a TODO comment and I create an issue. I guess there are multiple such cases. We can fix them at some point.

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 6 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp]: Yes, fixed

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 7 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp]: Yes, the same as above

File: src/smt/theory_str_noodler/formula_preprocess.cpp (hunk 8 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/formula_preprocess.cpp:1403]: I added a comment. Basically, we don't need this constraint anymore since it can be reconstructed from the newly added one.

**Comment by vhavlena** [comment]
Can I merge it?

File: src/smt/theory_str_noodler/formula_preprocess.cpp (hunk 8 above)
Comment (jurajsic) [reply src/smt/theory_str_noodler/formula_preprocess.cpp:1403]: Hmm, maybe it would be better if there were more comments in the function, at least connecting it with the explanation at the beginning. I assume `pr1` is the equation `X = Y W1 W2`, `pr2` is the second equation...  etc. And `flatten_concat` does something so we can use the third equation `K = W1 W2`?

Comment (vhavlena) [reply src/smt/theory_str_noodler/formula_preprocess.cpp:1419]: Ok, I added more comments.
